Userland/Libraries/LibVideo/VP9/Decoder.cpp

```diff
@@ -403,6 +406,745 @@ bool Decoder::trailing_bits()
+            auto& mv_hp_prob = m_probability_tables->mv_hp_prob();
+            mv_class0_hp_prob[i] = update_mv_prob(mv_class0_hp_prob[i]);
+            mv_hp_prob[i] = update_mv_prob(mv_hp_prob[i]);
+        }
+    }
+
+    return true;
+}
+
+u8 Decoder::update_mv_prob(u8 prob)
+{
+    if (m_bit_stream->read_bool(252)) {
+        return (m_bit_stream->read_literal(7) << 1u) | 1u;
+    }
+    return prob;
+}
+
+bool Decoder::setup_compound_reference_mode()
+{
+    if (m_ref_frame_sign_bias[LastFrame] == m_ref_frame_sign_bias[GoldenFrame]) {
+        m_comp_fixed_ref = AltRefFrame;
+        m_comp_var_ref[0] = LastFrame;
+        m_comp_var_ref[1] = GoldenFrame;
+    } else if (m_ref_frame_sign_bias[LastFrame] == m_ref_frame_sign_bias[AltRefFrame]) {
+        m_comp_fixed_ref = GoldenFrame;
+        m_comp_var_ref[0] = LastFrame;
+        m_comp_var_ref[1] = AltRefFrame;
+    } else {
+        m_comp_fixed_ref = LastFrame;
+        m_comp_var_ref[0] = GoldenFrame;
+        m_comp_var_ref[1] = AltRefFrame;
+    }
+    return true;
+}
+
+bool Decoder::decode_tiles()
+{
+    auto tile_cols = 1 << m_tile_cols_log2;
+    auto tile_rows = 1 << m_tile_rows_log2;
+    SAFE_CALL(clear_above_context());
+    for (auto tile_row = 0; tile_row < tile_rows; tile_row++) {
+        for (auto tile_col = 0; tile_col < tile_cols; tile_col++) {
+            auto last_tile = (tile_row == tile_rows - 1) && (tile_col == tile_cols - 1);
+            // FIXME: Spec has `sz -= tile_size + 4`, but I think we don't need this because our bit stream manages how much data we have left?
+            auto tile_size = last_tile ? m_bit_stream->bytes_remaining() : m_bit_stream->read_f(32);
+            m_mi_row_start = get_tile_offset(tile_row, m_mi_rows, m_tile_rows_log2);
+            m_mi_row_end = get_tile_offset(tile_row + 1, m_mi_rows, m_tile_rows_log2);
+            m_mi_col_start = get_tile_offset(tile_col, m_mi_cols, m_tile_cols_log2);
+            m_mi_col_end = get_tile_offset(tile_col + 1, m_mi_cols, m_tile_cols_log2);
+            SAFE_CALL(m_bit_stream->init_bool(tile_size));
+            SAFE_CALL(decode_tile());
+            SAFE_CALL(m_bit_stream->exit_bool());
+        }
+    }
+
+    return true;
+}
+
+bool Decoder::clear_above_context()
+{
+    if (!m_above_nonzero_context) {
+        m_above_nonzero_context = static_cast<u8**>(malloc(sizeof(u8) * 3 * m_mi_cols * 2));
+    } else {
+        __builtin_memset(m_above_nonzero_context, 0, sizeof(u8) * 3 * m_mi_cols * 2);
+    }
+
+    if (!m_above_seg_pred_context) {
+        m_above_seg_pred_context = static_cast<u8*>(malloc(sizeof(u8) * m_mi_cols));
+    } else {
+        __builtin_memset(m_above_seg_pred_context, 0, sizeof(u8) * m_mi_cols);
+    }
+
+    if (!m_above_partition_context) {
+        m_above_partition_context = static_cast<u8*>(malloc(sizeof(u8) * m_sb64_cols * 8));
+    } else {
+        __builtin_memset(m_above_partition_context, 0, sizeof(u8) * m_sb64_cols * 8);
+    }
```

Comment:
This could exists enough times to warrant a helper for it. Also, it don't understand why there is `u8**` in one case, but I don't understand this enough to tell you if this is wrong or not. :^)

Reply:
Yeah, this is defined in section 7.4.1 of the spec, the first of the three is a `u8[][]`, so this should be correct.